### DISCUSSION TITLE
Allow a configurable keep alive timeout in next config

### DISCUSSION THREAD

**Discussion by jordie23** [vercel/next.js#22211] [Ideas]
**Describe the feature you'd like to request**

NodeJS defaults `keepAliveTimeout` to 5 seconds. When NextJS is behind an AWS ELB/ALB, it can cause occasional 502 errors as the ELB's timeout is longer than Next's. To prevent this it should be the other way around where Next has a longer timeout than the ELB. Also need to be able to configure `headersTimeout` for this circumstance.

More info can be found here: https://shuheikagawa.com/blog/2019/04/25/keep-alive-timeout/

**Describe the solution you'd like**

Ability to set in the NextJS config a way to set the values for `keepAliveTimeout` and `headersTimeout`.

**Describe alternatives you've considered**

The only solution I was able to find was to modify the IdleTimeout on the AWS ELB to be 4 seconds, i.e 1 second less than node's default 5 seconds. This is a problem for any longer running requests (think 99th percentile or some intensive /api request) which will result in a 504 (gateway timeout). 

A custom server could also be set up however I don't think this is a unique use case (anyone hosting on AWS behind a load balancer will have this issue).

**Comment by narze** [comment]
@jordie23 I encounter this exact issue, did you solve it somehow?

**Comment by anuragkalia** [comment]
I will also want this feature, +1

**Comment by rawpixel-vincent** [comment]
another discussions about this https://github.com/vercel/next.js/discussions/19618

**Reply by rawpixel-vincent** [reply]
More information, reducing the ALB timeout to 4 seconds has a performance hit, implementing a custom server is a solution but I believe exposing this configuration with some side notes about the issue with load balancing will have a better impact on the next.js community overall.

**Comment by tabuchid** [comment]
How does anyone deploy next for anything serious outside of vercel? Between this and no way of enabling logging or integrating APM its very hard to maintain.

**Comment by danroestorf** [comment]
This is a serious issue on our application and setting the ALB timeout to 4s is not a viable option. NextJS is turning out to be a real headache in a production environment.

**Comment by jhchill666** [comment]
This was moved to a discussion, but doesn't seem to be discussed.

This is a serious problem for running NextJS in production, outside of Vercel.  While Next's commercial interests may lie with Vercel, this totally invalidates using NextJS in production on AWS, as a viable option.

Considering the fix is as simple as allowing some additional config on next's httpServer, can we not get this moving at all?

**Reply by tabuchid** [reply]
Another similar discussion here with similar response https://github.com/vercel/next.js/discussions/28642

**Comment by jhchill666** [comment]
I'm not too familiar with the internals of `next.config.js`, to provide additional configuration options to the underlying `httpServer`, but am happy to take this on

**Reply by tabuchid** [reply]
@timneutkens ^

**Comment by kalinin-egor** [comment]
Воу, очень круто!)

**Reply by jhchill666** [reply]
Keeping the discussion to english might be useful
